File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
// TC- O(n)
// SC - O(n)

/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
private:
    unordered_map<Node*, Node*> visited;
public:
    Node* copyRandomList(Node* head) {
        if(head==NULL)
            return head;
        
        if(visited.count(head))
            return visited[head];
        
        Node* curr = new Node(head->val, NULL, NULL);
        visited[head] = curr;
        curr->next = copyRandomList(head->next);
        curr->random = copyRandomList(head->random);
        return curr;
    }
};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""
# TC- O(n)
# SC- O(n)
class Solution:
    def __init__(self):
        self.mapping = {}

    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if head is None:
            return None

        if head in self.mapping:
            return self.mapping[head]

        curr = ListNode(head.val)
        self.mapping[head] = curr
        curr.next = self.copyRandomList(head.next)
        curr.random = self.copyRandomList(head.random)
        
        return curr